crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -367,6 +431,7 @@ where
                         let offset = FarmerCacheOffset::new(cache_index, piece_offset);
                         match maybe_piece_index {
                             Some(piece_index) => {
+                                *used_capacity = piece_offset.0 + 1;
```

Comment:
This naive logic is actually not correct. See, you get `maybe_piece_index`, there is nothing in API that guarantees that used pieces offsets do not have unused offsets in between them. In fact this is exactly what dangling offsets are supposed to represent: free offsets dangling inside of otherwise used capacity.

So used capacity is the whole range of used indices even if there are "holes" in it, while this is only increasing index for non-dangling offsets, which will result in inconsistent data structure.

Reply:
Yes, so what's actually being done here is to record the last_used_offset, as well as all free_offsets for the current cache_backend (including dangling_free_offset, you can see line 439, just below. ). And later on, we'll unify to compressing them.

Reply:
Hm... I see. This does work correctly, though it will potentially allocate a large `cache_stored_pieces` that would otherwise be empty or near-empty.